### ISSUE TITLE
Query compilation fails with an error about Timestamp for a query/data set with no timestamps

### ISSUE THREAD

**Issue by garyb** [quasar-analytics/quasar#2181]
This occurs on the old Mongo connector.

With this test data:

``` json
{ "b": { "ba": true, "bb": ["x", false, { "foo": "bar" }], "bc": 0.5 } }
{ "b": [1, 2, 3] }
{ "b": "test flat value" }
```

And this query:

``` sql 
SELECT
    CASE TYPE_OF(t.b)
    WHEN "map" THEN t.b{*}
    WHEN "array" THEN LENGTH(t.b)
    ELSE t.b END as b
FROM
    `/test/flatten_test.json` as t
``` 

The resulting error is:

``` json
{
  "error": {
    "status": "Unable to compile to JavaScript.",
    "detail": {
      "message": "Operation/value could not be compiled to JavaScript: Timestamp(1901-12-13T20:45:52Z, 0)",
      "value": "Timestamp(1901-12-13T20:45:52Z, 0)",
      "logicalplan": {
        "type": "LogicalPlan/Let",
        "label": "'__tmp0",
        "children": [
          {
            "type": "LogicalPlan/Let",
            "label": "'__tmp1",
            "children": [
              {
                "type": "LogicalPlan/Read",
                "label": "/test/flatten_test.json"
              },
              {
                "type": "LogicalPlan/Typecheck",
                "label": "Obj(Map[], Some(Top))",
                "children": [
                  {
                    "type": "LogicalPlan/Free",
                    "label": "'__tmp1"
                  },
                  {
                    "type": "LogicalPlan/Free",
                    "label": "'__tmp1"
                  },
                  {
                    "type": "LogicalPlan/Constant",
                    "label": "NA"
                  }
                ]
              }
            ]
          },
          {
            "type": "LogicalPlan/Invoke",
            "label": "Squash",
            "children": [
              {
                "type": "LogicalPlan/Invoke",
                "label": "MakeObject",
                "children": [
                  {
                    "type": "LogicalPlan/Constant/Str",
                    "label": "\"b\""
                  },
                  {
                    "type": "LogicalPlan/Let",
                    "label": "'__tmp2",
                    "children": [
                      {
                        "type": "LogicalPlan/ObjectProject",
                        "label": "Free('__tmp0){\"b\"}"
                      },
                      {
                        "type": "LogicalPlan/Typecheck",
                        "label": "Str",
                        "children": [
                          {
                            "type": "LogicalPlan/Free",
                            "label": "'__tmp2"
                          },
                          {
                            "type": "LogicalPlan/Invoke",
                            "label": "Cond",
                            "children": [
                              {
                                "type": "LogicalPlan/Invoke",
                                "label": "Eq",
                                "children": [
                                  {
                                    "type": "LogicalPlan/Invoke",
                                    "label": "unknown function: Returns the simple type of a value.",
                                    "children": [
                                      {
                                        "type": "LogicalPlan/ObjectProject",
                                        "label": "Free('__tmp0){\"b\"}"
                                      }
                                    ]
                                  },
                                  {
                                    "type": "LogicalPlan/Constant/Str",
                                    "label": "\"map\""
                                  }
                                ]
                              },
                              {
                                "type": "LogicalPlan/Invoke",
                                "label": "FlattenMap",
                                "children": [
                                  {
                                    "type": "LogicalPlan/Let",
                                    "label": "'__tmp3",
                                    "children": [
                                      {
                                        "type": "LogicalPlan/ObjectProject",
                                        "label": "Free('__tmp0){\"b\"}"
                                      },
                                      {
                                        "type": "LogicalPlan/Typecheck",
                                        "label": "Obj(Map[], Some(Top))",
                                        "children": [
                                          {
                                            "type": "LogicalPlan/Free",
                                            "label": "'__tmp3"
                                          },
                                          {
                                            "type": "LogicalPlan/Free",
                                            "label": "'__tmp3"
                                          },
                                          {
                                            "type": "LogicalPlan/Constant",
                                            "label": "Obj(Map( -> NA))"
                                          }
                                        ]
                                      }
                                    ]
                                  }
                                ]
                              },
                              {
                                "type": "LogicalPlan/Invoke",
                                "label": "Cond",
                                "children": [
                                  {
                                    "type": "LogicalPlan/Invoke",
                                    "label": "Eq",
                                    "children": [
                                      {
                                        "type": "LogicalPlan/Invoke",
                                        "label": "unknown function: Returns the simple type of a value.",
                                        "children": [
                                          {
                                            "type": "LogicalPlan/ObjectProject",
                                            "label": "Free('__tmp0){\"b\"}"
                                          }
                                        ]
                                      },
                                      {
                                        "type": "LogicalPlan/Constant/Str",
                                        "label": "\"array\""
                                      }
                                    ]
                                  },
                                  {
                                    "type": "LogicalPlan/Invoke",
                                    "label": "Length",
                                    "children": [
                                      {
                                        "type": "LogicalPlan/Free",
                                        "label": "'__tmp2"
                                      }
                                    ]
                                  },
                                  {
                                    "type": "LogicalPlan/ObjectProject",
                                    "label": "Free('__tmp0){\"b\"}"
                                  }
                                ]
                              }
                            ]
                          },
                          {
                            "type": "LogicalPlan/Constant",
                            "label": "NA"
                          }
                        ]
                      }
                    ]
                  }
                ]
              }
            ]
          }
        ]
      }
    }
  }
}
```

**Comment by sellout** [comment]
I think the Timestamp is showing up because the way we do typechecks in Mongo Expressions (since it has no way to actually check types) is abusing its total ordering of values. So we generate a bunch of values at the boundaries of different types and see how the value you’re checking compares to them.

The problem that shows up here is that we sometimes need to convert Mongo Expressions into JavaScript, and that is only a partial function, and the semantics don’t line up great (e.g., we have to identify when we’re using comparisons to fake type checking and convert it to actual type checks in JS).

Thankfully, I just yesterday committed a PR against the new Mongo branch that eliminates all Expression => JS conversions. Instead, we now have `Ior[Js, ExprOp]` as a result from compiling MapFunc (it used to be `Js \/ ExprOp`), so if we ever need to “convert” to JS we just grab the Js element from the Ior rather than trying to recover it from the Expression.

**Comment by zeeiyerWork** [comment]
We just need to verify once newMaster is available.

**Comment by drostron** [comment]
The query compiles without error against `next-major`.

**Comment by drostron** [comment]
Fixed via #2730
